Empty spanset fix

mdisibio opened the pull request:
**What this PR does**:
Changes how empty spansets `{ }` are handled to be more correct.  Instead of relying on the encoding vparquet1/2/3 to determine when no conditions exist, the engine pushes something down.   ~Marked as draft because want to verify the core behavior change first.~

Confirmed that the queries mentioned in the issue now work:
`{} | by(span.does_not_exist)`
`{} || { span.does_not_exist }`

~The tests for this query are failing which I think is expected:  `{ } | by(span.foo) | count() > 2`   The previous implied `span.foo exists` check is gone -- now a spanset will be created for the `nil` value (spans that matched `{ }` but didn't have that attribute).~

**Which issue(s) this PR fixes**:
Fixes #2498 

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

joe-elliott commented on pkg/traceql/ast_conditions.go:
```diff
@@ -2,6 +2,23 @@ package traceql
+	// For empty spansets { } ensure there is something that matches all spans.
+	// Use duration which would have been selected as part of the second pass
+	// metadata.
+	if s, ok := f.Expression.(Static); ok && s.Type == TypeBoolean && s.B {
```
"kind" is the column we scan when we just want to count spans. thoughts on using kind over duration?

there's weird edge cases where it would be better:
```
{} | { <needle in the haystack conditions> }
```

but obviously that's a silly query. is the rationale that any "sane" query with `{}` will return all spans? so just pull a metadata column we know we are going to get anyway?

final Q: do you think this is the longterm fix or we'd eventually want an Op that just force pulls all spans and let the fetch layer figure it out?

mdisibio replied:
I tried kind first.  The main drawback it shows up in the results unexpectedly. It was very noticeable in the UI and didn't seem right.   Not returning it also seemed like it would be tricky, the engine would have to keep track of when it was substituted, but not also not remove it was explicitly queried by another expression.

> want an Op that just force pulls all spans
I think the direction is to try to reduce specialty in the storage layer.  The recent changes to call out the metadata fields in the second pass is a good example of that.  Would like to see more of that, so initially I would say no to a special Op to return all spans.

joe-elliott reviewed (approved):
just a couple Q's, but approved.

if you get a chance can you double check benchmarks to make sure there's no unexpected side effects.

mdisibio commented:
Benchmarks.  Added 2 new ones.   Switched to StartTime, which ended up being nearly as efficient as kind, and also more universal since vParquet1 doesn't have a dedicated duration column.  Tweaked the spanset operation so that it remains fast and doesn't trigger this new logic (it detects there are already the desc,child,sibling intrinsics).  Finally, there is a slight decrease in I/O because we go from id+start+dur+kind to just the first three.

```
name                                                 old time/op    new time/op    delta
BackendBlockTraceQL/spanAttNameNoMatch-12              3.59ms ± 0%    3.68ms ± 6%    ~     (p=0.101 n=8+10)
BackendBlockTraceQL/spanAttValNoMatch-12               33.6ms ± 3%    34.3ms ± 5%    ~     (p=0.095 n=10+9)
BackendBlockTraceQL/spanAttValMatch-12                 39.2ms ± 0%    39.6ms ± 1%  +0.87%  (p=0.001 n=9+9)
BackendBlockTraceQL/spanAttIntrinsicNoMatch-12         3.61ms ± 2%    3.62ms ± 2%    ~     (p=0.460 n=10+8)
BackendBlockTraceQL/spanAttIntrinsicMatch-12           70.1ms ± 2%    70.5ms ± 1%    ~     (p=0.222 n=9+9)
BackendBlockTraceQL/spanAttIntrinsicRegexNoMatch-12    4.00ms ± 2%    3.96ms ± 1%  -1.12%  (p=0.019 n=10+10)
BackendBlockTraceQL/spanAttIntrinsicRegexMatch-12       106ms ± 1%     106ms ± 1%    ~     (p=0.393 n=10+10)
BackendBlockTraceQL/resourceAttNameNoMatch-12          2.96ms ± 4%    2.96ms ± 1%    ~     (p=0.423 n=9+8)
BackendBlockTraceQL/resourceAttValNoMatch-12           2.80ms ± 2%    2.79ms ± 0%    ~     (p=0.661 n=10+9)
BackendBlockTraceQL/resourceAttValMatch-12              513ms ± 1%     513ms ± 1%    ~     (p=0.743 n=9+8)
BackendBlockTraceQL/resourceAttIntrinsicNoMatch-12     2.82ms ± 2%    2.81ms ± 2%    ~     (p=0.460 n=10+8)
BackendBlockTraceQL/resourceAttIntrinsicMatch-12       70.4ms ± 3%    71.7ms ± 1%  +1.85%  (p=0.035 n=10+9)
BackendBlockTraceQL/mixedNameNoMatch-12                 708ms ± 2%     709ms ± 0%    ~     (p=0.156 n=10+9)
BackendBlockTraceQL/mixedValNoMatch-12                  676ms ± 1%     678ms ± 1%    ~     (p=0.211 n=9+10)
BackendBlockTraceQL/mixedValMixedMatchAnd-12           3.01ms ± 3%    3.02ms ± 4%    ~     (p=0.720 n=10+9)
BackendBlockTraceQL/mixedValMixedMatchOr-12             575ms ± 1%     576ms ± 1%    ~     (p=0.631 n=10+10)
BackendBlockTraceQL/mixedValBothMatch-12               2.85ms ± 2%    2.81ms ± 1%  -1.56%  (p=0.003 n=10+9)
BackendBlockTraceQL/count-12                            1.92s ± 4%     2.03s ± 2%  +5.78%  (p=0.000 n=9+9)
BackendBlockTraceQL/desc-12                             3.04s ±11%     2.89s ± 1%  -4.88%  (p=0.007 n=10+10)

name                                                 old speed      new speed      delta
BackendBlockTraceQL/spanAttNameNoMatch-12             894MB/s ± 0%   875MB/s ± 6%    ~     (p=0.101 n=8+10)
BackendBlockTraceQL/spanAttValNoMatch-12              230MB/s ± 3%   225MB/s ± 5%    ~     (p=0.095 n=10+9)
BackendBlockTraceQL/spanAttValMatch-12               95.1MB/s ± 0%  94.3MB/s ± 1%  -0.86%  (p=0.001 n=9+9)
BackendBlockTraceQL/spanAttIntrinsicNoMatch-12        812MB/s ± 2%   808MB/s ± 2%    ~     (p=0.460 n=10+8)
BackendBlockTraceQL/spanAttIntrinsicMatch-12          317MB/s ± 3%   316MB/s ± 1%    ~     (p=0.447 n=10+9)
BackendBlockTraceQL/spanAttIntrinsicRegexNoMatch-12   731MB/s ± 2%   739MB/s ± 1%  +1.12%  (p=0.019 n=10+10)
BackendBlockTraceQL/spanAttIntrinsicRegexMatch-12     210MB/s ± 1%   211MB/s ± 1%    ~     (p=0.393 n=10+10)
BackendBlockTraceQL/resourceAttNameNoMatch-12         427MB/s ± 4%   426MB/s ± 1%    ~     (p=0.423 n=9+8)
BackendBlockTraceQL/resourceAttValNoMatch-12          190MB/s ± 2%   191MB/s ± 0%    ~     (p=0.661 n=10+9)
BackendBlockTraceQL/resourceAttValMatch-12           76.0MB/s ± 1%  76.1MB/s ± 1%    ~     (p=0.743 n=9+8)
BackendBlockTraceQL/resourceAttIntrinsicNoMatch-12    238MB/s ± 2%   239MB/s ± 2%    ~     (p=0.460 n=10+8)
BackendBlockTraceQL/resourceAttIntrinsicMatch-12      518MB/s ± 3%   509MB/s ± 1%  -1.84%  (p=0.035 n=10+9)
BackendBlockTraceQL/mixedNameNoMatch-12              18.5MB/s ± 2%  18.4MB/s ± 0%    ~     (p=0.150 n=10+9)
BackendBlockTraceQL/mixedValNoMatch-12               19.3MB/s ± 1%  19.3MB/s ± 1%    ~     (p=0.190 n=9+10)
BackendBlockTraceQL/mixedValMixedMatchAnd-12          420MB/s ± 3%   418MB/s ± 4%    ~     (p=0.720 n=10+9)
BackendBlockTraceQL/mixedValMixedMatchOr-12          44.4MB/s ± 1%  44.3MB/s ± 1%    ~     (p=0.617 n=10+10)
BackendBlockTraceQL/mixedValBothMatch-12              236MB/s ± 2%   239MB/s ± 1%  +1.56%  (p=0.003 n=10+9)
BackendBlockTraceQL/count-12                         19.2MB/s ± 4%  17.6MB/s ± 2%  -8.32%  (p=0.000 n=9+9)
BackendBlockTraceQL/desc-12                          15.7MB/s ±11%  16.5MB/s ± 1%  +4.86%  (p=0.005 n=10+10)

name                                                 old MB_io/op   new MB_io/op   delta
BackendBlockTraceQL/spanAttNameNoMatch-12                3.21 ± 0%      3.21 ± 0%    ~     (all equal)
BackendBlockTraceQL/spanAttValNoMatch-12                 7.71 ± 0%      7.71 ± 0%    ~     (all equal)
BackendBlockTraceQL/spanAttValMatch-12                   3.73 ± 0%      3.73 ± 0%    ~     (all equal)
BackendBlockTraceQL/spanAttIntrinsicNoMatch-12           2.93 ± 0%      2.93 ± 0%    ~     (all equal)
BackendBlockTraceQL/spanAttIntrinsicMatch-12             22.3 ± 0%      22.3 ± 0%    ~     (all equal)
BackendBlockTraceQL/spanAttIntrinsicRegexNoMatch-12      2.93 ± 0%      2.93 ± 0%    ~     (all equal)
BackendBlockTraceQL/spanAttIntrinsicRegexMatch-12        22.3 ± 0%      22.3 ± 0%    ~     (all equal)
BackendBlockTraceQL/resourceAttNameNoMatch-12            1.26 ± 0%      1.26 ± 0%    ~     (all equal)
BackendBlockTraceQL/resourceAttValNoMatch-12             0.53 ± 0%      0.53 ± 0%    ~     (all equal)
BackendBlockTraceQL/resourceAttValMatch-12               39.0 ± 0%      39.0 ± 0%    ~     (all equal)
BackendBlockTraceQL/resourceAttIntrinsicNoMatch-12       0.67 ± 0%      0.67 ± 0%    ~     (all equal)
BackendBlockTraceQL/resourceAttIntrinsicMatch-12         36.5 ± 0%      36.5 ± 0%    ~     (all equal)
BackendBlockTraceQL/mixedNameNoMatch-12                  13.1 ± 0%      13.1 ± 0%    ~     (all equal)
BackendBlockTraceQL/mixedValNoMatch-12                   13.1 ± 0%      13.1 ± 0%    ~     (all equal)
BackendBlockTraceQL/mixedValMixedMatchAnd-12             1.26 ± 0%      1.26 ± 0%    ~     (all equal)
BackendBlockTraceQL/mixedValMixedMatchOr-12              25.5 ± 0%      25.5 ± 0%    ~     (all equal)
BackendBlockTraceQL/mixedValBothMatch-12                 0.67 ± 0%      0.67 ± 0%    ~     (all equal)
BackendBlockTraceQL/count-12                             36.9 ± 0%      35.7 ± 0%  -3.01%  (p=0.000 n=10+10)
BackendBlockTraceQL/desc-12                              47.7 ± 0%      47.7 ± 0%    ~     (all equal)

name                                                 old alloc/op   new alloc/op   delta
BackendBlockTraceQL/spanAttNameNoMatch-12              1.44MB ± 3%    1.42MB ± 1%    ~     (p=0.065 n=10+9)
BackendBlockTraceQL/spanAttValNoMatch-12                101MB ± 2%     102MB ± 2%    ~     (p=0.190 n=9+9)
BackendBlockTraceQL/spanAttValMatch-12                 1.44MB ± 7%    1.49MB ± 8%    ~     (p=0.102 n=10+10)
BackendBlockTraceQL/spanAttIntrinsicNoMatch-12         1.70MB ± 2%    1.67MB ± 1%  -1.45%  (p=0.000 n=10+9)
BackendBlockTraceQL/spanAttIntrinsicMatch-12           4.04MB ± 1%    4.18MB ± 8%    ~     (p=0.315 n=7+10)
BackendBlockTraceQL/spanAttIntrinsicRegexNoMatch-12    1.91MB ± 3%    1.89MB ± 2%    ~     (p=0.123 n=10+10)
BackendBlockTraceQL/spanAttIntrinsicRegexMatch-12      8.35MB ± 1%    8.36MB ± 5%    ~     (p=0.696 n=8+10)
BackendBlockTraceQL/resourceAttNameNoMatch-12          1.35MB ± 3%    1.35MB ± 3%    ~     (p=0.393 n=10+10)
BackendBlockTraceQL/resourceAttValNoMatch-12           1.33MB ± 2%    1.33MB ± 2%    ~     (p=0.842 n=9+10)
BackendBlockTraceQL/resourceAttValMatch-12              186MB ± 1%     185MB ± 0%  -0.48%  (p=0.009 n=10+10)
BackendBlockTraceQL/resourceAttIntrinsicNoMatch-12     1.34MB ± 2%    1.34MB ± 1%    ~     (p=1.000 n=10+9)
BackendBlockTraceQL/resourceAttIntrinsicMatch-12       2.57MB ±12%    2.45MB ±13%    ~     (p=0.165 n=10+10)
BackendBlockTraceQL/mixedNameNoMatch-12                 111MB ± 1%     110MB ± 1%  -0.73%  (p=0.003 n=9+8)
BackendBlockTraceQL/mixedValNoMatch-12                  110MB ± 1%     110MB ± 1%    ~     (p=0.881 n=9+9)
BackendBlockTraceQL/mixedValMixedMatchAnd-12           1.35MB ± 3%    1.35MB ± 3%    ~     (p=0.631 n=10+10)
BackendBlockTraceQL/mixedValMixedMatchOr-12            4.53MB ±41%    4.60MB ±21%    ~     (p=0.436 n=10+10)
BackendBlockTraceQL/mixedValBothMatch-12               1.34MB ± 2%    1.35MB ± 2%    ~     (p=0.393 n=10+10)
BackendBlockTraceQL/count-12                            939MB ± 0%     940MB ± 0%    ~     (p=0.400 n=9+10)
BackendBlockTraceQL/desc-12                             693MB ± 0%     693MB ± 0%    ~     (p=1.000 n=10+9)

name                                                 old allocs/op  new allocs/op  delta
BackendBlockTraceQL/spanAttNameNoMatch-12               18.7k ± 0%     18.7k ± 0%    ~     (p=0.173 n=10+9)
BackendBlockTraceQL/spanAttValNoMatch-12                19.0k ± 0%     19.0k ± 0%    ~     (p=0.380 n=10+8)
BackendBlockTraceQL/spanAttValMatch-12                  19.1k ± 0%     19.1k ± 0%    ~     (p=0.372 n=10+10)
BackendBlockTraceQL/spanAttIntrinsicNoMatch-12          18.7k ± 0%     18.7k ± 0%  -0.00%  (p=0.022 n=9+10)
BackendBlockTraceQL/spanAttIntrinsicMatch-12            30.2k ± 0%     30.2k ± 0%    ~     (p=0.212 n=10+10)
BackendBlockTraceQL/spanAttIntrinsicRegexNoMatch-12     21.0k ± 0%     21.0k ± 0%    ~     (p=0.465 n=10+10)
BackendBlockTraceQL/spanAttIntrinsicRegexMatch-12       73.0k ± 0%     73.0k ± 0%    ~     (p=0.867 n=10+10)
BackendBlockTraceQL/resourceAttNameNoMatch-12           18.7k ± 0%     18.7k ± 0%    ~     (all equal)
BackendBlockTraceQL/resourceAttValNoMatch-12            18.7k ± 0%     18.7k ± 0%    ~     (p=1.000 n=10+10)
BackendBlockTraceQL/resourceAttValMatch-12              1.69M ± 0%     1.69M ± 0%  -0.00%  (p=0.037 n=9+10)
BackendBlockTraceQL/resourceAttIntrinsicNoMatch-12      18.7k ± 0%     18.7k ± 0%    ~     (all equal)
BackendBlockTraceQL/resourceAttIntrinsicMatch-12        24.0k ± 0%     24.0k ± 0%  -0.02%  (p=0.027 n=10+10)
BackendBlockTraceQL/mixedNameNoMatch-12                 21.1k ± 0%     21.1k ± 0%  -0.03%  (p=0.003 n=10+9)
BackendBlockTraceQL/mixedValNoMatch-12                  21.9k ± 0%     21.9k ± 0%    ~     (p=0.421 n=8+8)
BackendBlockTraceQL/mixedValMixedMatchAnd-12            18.7k ± 0%     18.7k ± 0%    ~     (all equal)
BackendBlockTraceQL/mixedValMixedMatchOr-12             31.2k ± 0%     31.2k ± 0%    ~     (p=0.344 n=9+10)
BackendBlockTraceQL/mixedValBothMatch-12                18.7k ± 0%     18.7k ± 0%    ~     (p=0.248 n=10+9)
BackendBlockTraceQL/count-12                            10.1M ± 0%     10.1M ± 0%  -0.00%  (p=0.000 n=10+10)
BackendBlockTraceQL/desc-12                             6.56M ± 0%     6.56M ± 0%    ~     (p=0.497 n=9+10)
```
